src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -119,139 +124,172 @@ impl NailgunPool {
                 Self::start_new_nailgun(
-                  &mut *processes,
+                  processes,
                   name,
                   startup_options,
-                  &workdir_path,
+                  workdir_path,
                   requested_server_fingerprint,
                   build_id_requesting_connection,
-                )
+                  store,
+                  input_files,
+                  workunit_store,
+                  jdk_path,
+                ).to_boxed()
               }
+              Err(e) => future::err(e).to_boxed(),
             }
-          }
-          Ok(_) => {
-            // The process has exited with some exit code
-            debug!("The requested nailgun server was not running anymore. Restarting process...");
-            Self::start_new_nailgun(
-              &mut *processes,
-              name,
-              startup_options,
-              &workdir_path,
-              requested_server_fingerprint,
-              build_id_requesting_connection,
-            )
-          }
-          Err(e) => Err(e),
-        }
-      } else {
-        // We don't have a running nailgun registered in the map.
-        debug!(
-          "No nailgun server is running with name {}. Starting one...",
-          &name
-        );
-        Self::start_new_nailgun(
-          &mut *processes,
-          name,
-          startup_options,
-          &workdir_path,
-          requested_server_fingerprint,
-          build_id_requesting_connection,
-        )
-      };
-      debug!("Unlocking nailgun process pool.");
-      connection_result
-    })
-    .to_boxed()
+          }).to_boxed()
+        } else {
+          // We don't have a running nailgun registered in the map.
+          debug!(
+            "No nailgun server is running with name {}. Starting one...",
+            &name
+          );
+          Self::start_new_nailgun(
+            processes,
+            name,
+            startup_options,
+            workdir_path,
+            requested_server_fingerprint,
+            build_id_requesting_connection,
+            store,
+            input_files,
+            workunit_store,
+            jdk_path,
+          ).to_boxed()
+        };
+        debug!("Unlocking nailgun process pool.");
+        connection_result
+    }).to_boxed()
   }
 
-  //
-  // This is a blocking method that is being called under the NailgunProcessMap's lock (see #8543)
-  //
   fn start_new_nailgun(
-    processes: &mut NailgunProcessMap,
+    mut processes: MutexGuard<NailgunProcessMap>,
     name: String,
     startup_options: ExecuteProcessRequest,
-    workdir_path: &PathBuf,
+    workdir_path: PathBuf,
     nailgun_server_fingerprint: NailgunProcessFingerprint,
     build_id: String,
-  ) -> Result<Port, String> {
+    store: Store,
+    input_files: Digest,
+    workunit_store: WorkUnitStore,
+    jdk_path: PathBuf,
+  ) -> BoxFuture<(Port, MutexGuard<NailgunProcess>), String> {
```

Comment:
The `NailgunProcess` has a `port` field, so can probably skip the separate port return here.